br/pkg/rtree/rtree.go

```diff
@@ -293,3 +302,125 @@ type ProgressRange struct {
+	ta := than.(*ProgressRange)
+	return bytes.Compare(pr.Origin.StartKey, ta.Origin.StartKey) < 0
+}
+
+var _ btree.Item = &ProgressRange{}
+
+// ProgressRangeTree is a sorted tree for ProgressRanges.
+// All the progress ranges it sorted do not overlap.
+type ProgressRangeTree struct {
+	*btree.BTree
+}
+
+// NewProgressRangeTree returns an empty range tree.
+func NewProgressRangeTree() ProgressRangeTree {
+	return ProgressRangeTree{
+		BTree: btree.New(32),
+	}
+}
+
+// find is a helper function to find an item that contains the range.
+func (rangeTree *ProgressRangeTree) find(pr *ProgressRange) *ProgressRange {
+	var ret *ProgressRange
+	rangeTree.DescendLessOrEqual(pr, func(item btree.Item) bool {
+		ret = item.(*ProgressRange)
+		return false
+	})
+
+	if ret == nil || !ret.Origin.Contains(pr.Origin.StartKey) {
+		return nil
+	}
+
+	return ret
+}
+
+// Insert inserts a ProgressRange into the tree, it will return an error if there is a overlapping range.
+func (rangeTree *ProgressRangeTree) Insert(pr *ProgressRange) error {
+	overlap := rangeTree.find(pr)
+	if overlap != nil {
+		return errors.Errorf("failed to insert the progress range into range tree, "+
+			"because there is a overlapping range. The insert item start key: %s; "+
+			"The overlapped item start key: %s, end key: %s.",
+			redact.Key(pr.Origin.StartKey), redact.Key(overlap.Origin.StartKey), redact.Key(overlap.Origin.EndKey))
+	}
+	rangeTree.ReplaceOrInsert(pr)
+	return nil
+}
+
+// FindContained finds if there is a progress range containing the key range [startKey, endKey).
+func (rangeTree *ProgressRangeTree) FindContained(startKey, endKey []byte) (*ProgressRange, error) {
+	var ret *ProgressRange
+	rangeTree.Descend(func(item btree.Item) bool {
+		pr := item.(*ProgressRange)
+		if bytes.Compare(pr.Origin.StartKey, startKey) <= 0 {
+			ret = pr
+			return false
+		}
+		return true
+	})
+
+	if ret == nil {
+		return nil, errors.Errorf("Cannot find progress range that contains the start key: %s", redact.Key(startKey))
+	}
+
+	if !ret.Origin.ContainsRegion(startKey, endKey) {
+		return nil, errors.Errorf("The given region is not contained in the found progress range. "+
+			"The region start key is %s; The progress range start key is %s, end key is %s.",
+			startKey, redact.Key(ret.Origin.StartKey), redact.Key(ret.Origin.EndKey))
+	}
+
+	return ret, nil
+}
+
+type progressRangeIterItem struct {
+	pr       *ProgressRange
+	complete bool
+}
+
+type ProgressRangeIter struct {
+	items []*progressRangeIterItem
+	left  int
+}
+
+func (rangeTree *ProgressRangeTree) Iter() *ProgressRangeIter {
+	items := make([]*progressRangeIterItem, 0, rangeTree.Len())
+	rangeTree.Ascend(func(item btree.Item) bool {
+		items = append(items, &progressRangeIterItem{
+			pr:       item.(*ProgressRange),
+			complete: false,
+		})
+		return true
+	})
+	return &ProgressRangeIter{
+		items: items,
+		left:  len(items),
+	}
+}
+
+func (iter *ProgressRangeIter) GetIncompleteRanges() []Range {
```

Comment:
In fact this "iterator" yields all its content and can be "iterated" multi times. Maybe rename `ProgressRangeIter` to `IncompleteRangesFetcher` or something?